tools/traffic/workers/blob_reader.go

```diff
@@ -0,0 +1,223 @@
+	generatorMetrics metrics.Metrics) BlobReader {
+
+	return BlobReader{
+		ctx:                        ctx,
+		waitGroup:                  waitGroup,
+		logger:                     logger,
+		config:                     config,
+		retriever:                  retriever,
+		chainClient:                chainClient,
+		blobsToRead:                blobStore,
+		generatorMetrics:           generatorMetrics,
+		fetchBatchHeaderMetric:     generatorMetrics.NewLatencyMetric("fetch_batch_header"),
+		fetchBatchHeaderSuccess:    generatorMetrics.NewCountMetric("fetch_batch_header_success"),
+		fetchBatchHeaderFailure:    generatorMetrics.NewCountMetric("fetch_batch_header_failure"),
+		recombinationSuccessMetric: generatorMetrics.NewCountMetric("recombination_success"),
+		recombinationFailureMetric: generatorMetrics.NewCountMetric("recombination_failure"),
+		readLatencyMetric:          generatorMetrics.NewLatencyMetric("read"),
+		validBlobMetric:            generatorMetrics.NewCountMetric("valid_blob"),
+		invalidBlobMetric:          generatorMetrics.NewCountMetric("invalid_blob"),
+		readSuccessMetric:          generatorMetrics.NewCountMetric("read_success"),
+		readFailureMetric:          generatorMetrics.NewCountMetric("read_failure"),
+		operatorSuccessMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+		operatorFailureMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+		requiredReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("required_read_pool_size"),
+		optionalReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("optional_read_pool_size"),
+	}
+}
+
+// Start begins a blob reader goroutine.
+func (reader *BlobReader) Start() {
+	reader.waitGroup.Add(1)
+	ticker := time.NewTicker(reader.config.ReadRequestInterval)
+	go func() {
+		defer reader.waitGroup.Done()
+		for {
+			select {
+			case <-(*reader.ctx).Done():
+				return
+			case <-ticker.C:
+				reader.randomRead()
+			}
+		}
+	}()
+}
+
+// randomRead reads a random blob.
+func (reader *BlobReader) randomRead() {
+	metadata := reader.blobsToRead.GetNext()
+	if metadata == nil {
+		// There are no blobs that we are required to read.
+		return
+	}
+
+	reader.requiredReadPoolSizeMetric.Set(float64(reader.blobsToRead.Size()))
+
+	ctxTimeout, cancel := context.WithTimeout(*reader.ctx, reader.config.FetchBatchHeaderTimeout)
+	batchHeader, err := metrics.InvokeAndReportLatency(reader.fetchBatchHeaderMetric,
+		func() (*contractEigenDAServiceManager.IEigenDAServiceManagerBatchHeader, error) {
+			return reader.chainClient.FetchBatchHeader(
+				ctxTimeout,
+				gcommon.HexToAddress(reader.config.EigenDAServiceManager),
+				metadata.BatchHeaderHash[:],
+				big.NewInt(int64(0)),
+				nil)
+		})
+	cancel()
+	if err != nil {
+		reader.logger.Error("failed to get batch header", "err:", err)
+		reader.fetchBatchHeaderFailure.Increment()
+		return
+	}
+	reader.fetchBatchHeaderSuccess.Increment()
+
+	var batchHeaderHash [32]byte
+	copy(batchHeaderHash[:], metadata.BatchHeaderHash[:])
```

Comment:
why do we need this copy?

Reply:
Hmm... I honestly can't remember why I did it that way. Refactored so that it doesn't use a copy.